File: runme/mcgen/local/datan/datsrc/auxhes.f
      SUBROUTINE AUXHES(X,HESSE,N,NRED,LIST,FUNC)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      PARAMETER(DELTA=1.D-5,CUT=1.D-9)
      DIMENSION X(*),HESSE(NRED,*),LIST(*)
      EXTERNAL FUNC
      IF(NRED.GT.0) THEN
        IL=0
        DO 20 I=1,N
          KL=0
          IF(LIST(I).EQ. 1.OR. N.EQ.NRED) THEN
            IL=IL+1
            DO 10 K=1,N
              IF(LIST(K).EQ. 1.OR. NRED.EQ.N) THEN
                KL=KL+1
                IF(KL.LT.IL) THEN
                  HESSE(IL,KL)=HESSE(KL,IL)
                ELSE
                  E=ABS(X(I))
                  IF(E.LT.CUT) E=CUT
                  DELI=DELTA*E
                  SAVI=X(I)
                  IF (K.EQ.I) THEN
                    F=FUNC(X,N)
                    X(I)=X(I)+DELI
                    FP=FUNC(X,N)
                    X(I)=SAVI-DELI
                    FM=FUNC(X,N)
                    X(I)=SAVI
                    HESSE(IL,IL)=((FP-F)/DELI-(F-FM)/DELI)/DELI
                  ELSE
                    E=ABS(X(K))
                    IF(E.LT.CUT) E=CUT
                    DELK=DELTA*E
                    SAVK=X(K)
                    X(K)=SAVK+DELK
                    X(I)=SAVI+DELI
                    FP=FUNC(X,N)
                    X(I)=SAVI-DELI
                    FM=FUNC(X,N)
                    DFDXIP=(FP-FM)/(DELI+DELI)
                    X(K)=SAVK-DELK
                    X(I)=SAVI+DELI
                    FP=FUNC(X,N)
                    X(I)=SAVI-DELI
                    FM=FUNC(X,N)
                    X(I)=SAVI
                    X(K)=SAVK
                    DFDXIM=(FP-FM)/(DELI+DELI)
                    HESSE(IL,KL)=(DFDXIP-DFDXIM)/(DELK+DELK)
                  END IF
                END IF
              END IF
   10       CONTINUE
          END IF
   20   CONTINUE
      END IF
      END

File: runme/mcgen/local/datan/datsrc/auxhes.c
#include "datsrc.h"

void LIBFUNC auxhes_(double P_T *x, double P_T *hesse, integer const P_T *n,
             integer const P_T *nred, integer const P_T *list,
             double (CBFUNC P_T *func)(double const P_T *,
             integer const P_T *))
{
    /* System generated locals */
    integer hesse_dim1, hesse_offset, i__1, i__2;
    double d__1;

    /* Local variables */
    integer i, k, il, kl;
    double deli, delk, savi, savk, e, f, fm, fp, dfdxim, dfdxip;

    /* Parameter adjustments */
    --x;
    hesse_dim1 = *nred;
    hesse_offset = hesse_dim1 + 1;
    hesse -= hesse_offset;
    --list;

    /* Function Body */
    if (*nred > 0) {
        il = 0;
        i__1 = *n;
        for (i = 1; i <= i__1; ++i) {
            kl = 0;
            if (list[i] == 1 || *n == *nred) {
                ++il;
                i__2 = *n;
                for (k = 1; k <= i__2; ++k) {
                    if (list[k] == 1 || *nred == *n) {
                        ++kl;
                        if (kl < il) {
                            hesse[il + kl * hesse_dim1] = hesse[kl + il * 
                                    hesse_dim1];
                        } else {
                            e = (d__1 = x[i], abs(d__1));
                            if (e < 1e-9) {
                                e = 1e-9;
                            }
                            deli = e * 1e-5;
                            savi = x[i];
                            if (k == i) {
                                f = (*func)(&x[1], n);
                                x[i] += deli;
                                fp = (*func)(&x[1], n);
                                x[i] = savi - deli;
                                fm = (*func)(&x[1], n);
                                x[i] = savi;
                                hesse[il + il * hesse_dim1] = ((fp - f) / 
                                        deli - (f - fm) / deli) / deli;
                            } else {
                                e = (d__1 = x[k], abs(d__1));
                                if (e < 1e-9) {
                                    e = 1e-9;
                                }
                                delk = e * 1e-5;
                                savk = x[k];
                                x[k] = savk + delk;
                                x[i] = savi + deli;
                                fp = (*func)(&x[1], n);
                                x[i] = savi - deli;
                                fm = (*func)(&x[1], n);
                                dfdxip = (fp - fm) / (deli + deli);
                                x[k] = savk - delk;
                                x[i] = savi + deli;
                                fp = (*func)(&x[1], n);
                                x[i] = savi - deli;
                                fm = (*func)(&x[1], n);
                                x[i] = savi;
                                x[k] = savk;
                                dfdxim = (fp - fm) / (deli + deli);
                                hesse[il + kl * hesse_dim1] = (dfdxip - 
                                        dfdxim) / (delk + delk);
                            }
                        }
                    }
                }
            }
        }
    }
    return;
} /* auxhes_ */

